File: src/main/java/com/blackcowmoo/moomark/auth/service/UserService.java
package com.blackcowmoo.moomark.auth.service;

import com.blackcowmoo.moomark.auth.model.AuthProvider;
import com.blackcowmoo.moomark.auth.model.Role;
import com.blackcowmoo.moomark.auth.model.entity.User;
import com.blackcowmoo.moomark.auth.repository.UserRepository;

import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.stereotype.Service;

@Service
public class UserService {
  @Autowired
  UserRepository userRepository;

  public User getUserById(AuthProvider authProvider, String id) {
    return userRepository.findByIdAndAuthProvider(id, authProvider);
  }

  public User signUp(String id, AuthProvider authProvider, String nickname, String email, String picture) {
    return userRepository.save(new User(id, authProvider, email, nickname, picture, Role.USER));
  }

  public void withdraw(String id, AuthProvider authProvider) {
    userRepository.delete(userRepository.findByIdAndAuthProvider(id, authProvider));
  }

  public void withdraw(User user) {
    userRepository.delete(user);
  }

  public User updateUser(User user, String nickname, String picture) {
    user.updateNickname(nickname);
    user.updatePicture(picture);
    return userRepository.save(user);
  }

  // @Override
  // public boolean updateUserNickname(long userId, String nickname) {
  // try {
  // User willUpdateUser = userRepository.getOne(userId);
  // willUpdateUser.updateNickname(nickname);
  // updateUser(willUpdateUser);
  // } catch (Exception e) {
  // log.error(e);
  // return false;
  // }
  // return true;
  // }

}

File: src/main/kotlin/com/blackcowmoo/moomark/auth/service/UserService.kt
package com.blackcowmoo.moomark.auth.service

import com.blackcowmoo.moomark.auth.model.AuthProvider
import com.blackcowmoo.moomark.auth.model.Role
import com.blackcowmoo.moomark.auth.model.entity.User
import com.blackcowmoo.moomark.auth.repository.UserRepository
import org.springframework.beans.factory.annotation.Autowired
import org.springframework.beans.factory.annotation.Value
import org.springframework.stereotype.Service

@Service
open class UserService {

  @Value("\${resources.user.default-picture}")
  private lateinit var defaultPicture: String

  @Autowired
  private lateinit var userRepository: UserRepository

  open fun getUserById(authProvider: AuthProvider, id: String): User? {
    return userRepository.findByIdAndAuthProvider(id, authProvider)
  }

  open fun signUp(id: String, authProvider: AuthProvider, nickname: String, email: String, picture: String): User {
    return userRepository.save(User(id, authProvider, email, nickname, picture, Role.USER))
  }

  open fun withdraw(id: String, authProvider: AuthProvider) {
    userRepository.delete(userRepository.findByIdAndAuthProvider(id, authProvider))
  }

  open fun withdraw(user: User) {
    userRepository.delete(user)
  }

  open fun updateUser(user: User, nickname: String?, picture: String?): User {
    if (nickname != null && nickname.isNotEmpty()) {
      user.updateNickname(nickname)
    }

    if (picture != null && picture.isNotEmpty() && picture.startsWith("https://")) {
      user.updatePicture(picture)
    } else if (picture != null && picture.isEmpty() && nickname == null) {
      user.updatePicture(defaultPicture)
    }

    return userRepository.save(user)
  }
}
